qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -1403,6 +1403,98 @@ def num_basis_gates(self, unitary):
         return np.argmax([trace_to_fid(traces[i]) * self.basis_fidelity**i for i in range(4)])
 
 
+class TwoQubitDecomposeUpToDiagonal:
+    """
+    Class to decompose two qubit unitaries into the product of a diagonal gate
+    and another unitary gate which can be represented by two CX gates instead of the
+    usual three. This can be used when neighboring gates commute with the diagonal to
+    potentially reduce overall CX count.
+    """
+
+    def __init__(self):
+        self.sy = np.array([[0, -1j], [1j, 0]])
+        self.sz = np.array([[1, 0], [0, -1]])
+        self.sysy = np.kron(self.sy, self.sy)
+
+    def _u4_to_su4(self, u4):
+        from scipy import linalg as la
+
+        phase_factor = np.conj(la.det(u4) ** (-1 / u4.shape[0]))
+        su4 = u4 / phase_factor
+        return su4, cmath.phase(phase_factor)
+
+    def _gamma(self, mat):
+        """
+        proposition II.1: this invariant characterizes when two operators in U(4),
+        say u, v, are equivalent up to single qubit gates:
+
+           u ≡ v -> Det(γ(u)) = Det(±(γ(v)))
+        """
+        sumat, _ = self._u4_to_su4(mat)
+        sysy = self.sysy
+        return sumat @ sysy @ sumat.T @ sysy
+
+    def _cx0_test(self, mat):
+        # proposition III.1: zero cx sufficient
+        gamma = self._gamma(mat)
+        evals = np.linalg.eigvals(gamma)
+        return np.all(np.isclose(evals, np.ones(4)))
+
+    def _cx1_test(self, mat):
+        # proposition III.2: one cx sufficient
+        gamma = self._gamma(mat)
+        evals = np.linalg.eigvals(gamma)
+        uvals, ucnts = np.unique(np.round(evals, 10), return_counts=True)
+        return (
+            len(uvals) == 2
+            and all(ucnts == 2)
+            and all((np.isclose(x, 1j)) or np.isclose(x, -1j) for x in uvals)
+        )
+
+    def _cx2_test(self, mat):
+        # proposition III.3: two cx sufficient
+        gamma = self._gamma(mat)
+        return np.isclose(np.trace(gamma).imag, 0)
+
+    def _real_trace_transform(self, mat):
+        """
+        Determine diagonal gate such that
+
+        U3 = D U2
+
+        Where U3 is a general two-qubit gate which takes 3 cnots, D is a
+        diagonal gate, and U2 is a gate which takes 2 cnots.
+        """
+        a1 = (
+            -mat[1, 3] * mat[2, 0]
+            + mat[1, 2] * mat[2, 1]
+            + mat[1, 1] * mat[2, 2]
+            - mat[1, 0] * mat[2, 3]
+        )
+        a2 = (
+            mat[0, 3] * mat[3, 0]
+            - mat[0, 2] * mat[3, 1]
+            - mat[0, 1] * mat[3, 2]
+            + mat[0, 0] * mat[3, 3]
+        )
+        theta = 0  # arbitrary
+        phi = 0  # arbitrary
+        psi = np.arctan2(a1.imag + a2.imag, a1.real - a2.real) - phi
+        diag = np.diag(np.exp(-1j * np.array([theta, phi, psi, -(theta + phi + psi)])))
```

Comment:
I didn't want to eliminate theta and phi just yet. Perhaps in the future it may be useful to make these non-zero, although I don't have a concrete example now.